File: node_version/convertToTxt.js
const fs = require('fs');
const path = require('path');

function main() {
  // Modify the source directory path according to your needs
  const sourceDir = './test';

  // Modify the output directory path according to your needs
  const outputDir = './output';

  // Modify the file types you want to process
  const fileTypes = ['.md', '.js', '.ts', '.json', '.sql'];

  // Check if the output directory exists, if not, create it
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  // Walk through the source directory to find files with the specified file types
  const files = getFilesRecursive(sourceDir, fileTypes);

  // Process each file
  files.forEach((file) => {
    const inputFilePath = file;
    const relativePath = path.relative(sourceDir, inputFilePath);
    const newRoot = path.dirname(inputFilePath.replace(sourceDir, outputDir));
    const outputFile = path.join(
      newRoot,
      `${path.parse(file).name}.${path.parse(file).ext.slice(1)}.txt`
    );

    // Create the output directory if it doesn't exist
    if (!fs.existsSync(newRoot)) {
      fs.mkdirSync(newRoot, { recursive: true });
    }

    // Read the input file and write the output file with custom text
    const fileContent = fs.readFileSync(inputFilePath, 'utf8');
    const outputContent = `This is a txt representation of the VirtueMaster file located at ${relativePath}\n\n${fileContent}`;
    fs.writeFileSync(outputFile, outputContent, 'utf8');
  });

  // Process the README.md file separately
  const readmeFile = files.find((file) => path.basename(file) === 'README.md');
  if (readmeFile) {
    const readmeInputFilePath = readmeFile;
    const readmeRelativePath = path.relative(sourceDir, readmeInputFilePath);
    const readmeOutputFile = path.join(
      path.dirname(readmeInputFilePath.replace(sourceDir, outputDir)),
      `${path.parse('README.md').name}.${path.parse('README.md').ext.slice(1)}.txt`
    );

    // Read the README.md file and write the output file with custom text
    const readmeContent = fs.readFileSync(readmeInputFilePath, 'utf8');
    const outputContent = `This is a txt representation of the VirtueMaster file located at ${readmeRelativePath}\n\n${readmeContent}`;
    fs.writeFileSync(readmeOutputFile, outputContent, 'utf8');
  }
}

function getFilesRecursive(dir, fileTypes) {
  let files = [];

  if (!fs.existsSync(dir)) {
    return files;
  }

  const dirEntries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of dirEntries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      files = files.concat(getFilesRecursive(fullPath, fileTypes));
    } else if (entry.isFile() && fileTypes.includes(path.extname(fullPath))) {
      files.push(fullPath);
    }
  }

  return files;
}

main();

File: python_version/convertToTxt.py
import os
import shutil

def main():
    # Modify the source directory path according to your needs
    source_dir = './test'

    # Modify the output directory path according to your needs
    output_dir = './output'

    # Modify the file types you want to process
    file_types = ('.md', '.js', '.ts', '.json', '.sql')

    # Initialize the variable for the README.md file
    readme_file = None

    # Check if the output directory exists, if not, create it
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    # Walk through the source directory to find files with the specified file types
    for root, dirs, files in os.walk(source_dir):
        for file in files:
            if file.endswith(file_types) and file != 'README.md':
                # Create the input and output file paths
                input_file_path = os.path.join(root, file)
                relative_path = os.path.relpath(input_file_path, source_dir)
                new_root = root.replace(source_dir, output_dir)
                output_file_path = os.path.join(new_root, f'{os.path.splitext(file)[0]}.{os.path.splitext(file)[1][1:]}.txt')

                # Create the output directory if it doesn't exist
                if not os.path.exists(new_root):
                    os.makedirs(new_root)

                # Read the input file and write the output file with custom text
                with open(input_file_path, 'r', encoding='ISO-8859-1') as infile, open(output_file_path, 'w', encoding='ISO-8859-1') as outfile:
                    # Change the message here if needed
                    outfile.write(f'This is a txt representation of the VirtueMaster file located at {relative_path}\n\n')
                    outfile.write(infile.read())

            # Process the README.md file separately
            elif file == 'README.md':
                readme_input_file_path = os.path.join(root, file)
                readme_relative_path = os.path.relpath(readme_input_file_path, source_dir)

    # Convert the README.md file to a txt file
    if readme_file:
        readme_new_root = readme_input_file_path.replace(source_dir, output_dir)
        readme_output_file_path = os.path.join(os.path.dirname(readme_new_root), f'{os.path.splitext("README.md")[0]}.{os.path.splitext("README.md")[1][1:]}.txt')
        with open(readme_input_file_path, 'r', encoding='ISO-8859-1') as infile, open(readme_output_file_path, 'w', encoding='ISO-8859-1') as outfile:
            # Change the message here if needed
            outfile.write(f'This is a txt representation of the VirtueMaster file located at {readme_relative_path}\n\n')
            outfile.write(infile.read())

if __name__ == '__main__':
    main()